#6595: Provide input validation for Bumper length. Enhance QuickPromp…

svatoun opened the pull request:
…tUtil to display validation messages

I enhanced that entry with a validation + a simple to use extension to `QuickPromptUtil` API to validate integer inputs. The support can be used to extend validation to other data types produced from a String, if necessary.

---

Please note that I have used NetBeans Form editor to produce a UI panel; although the JMRI sources contains `nbproject` folder and even some NetBeans `.form` files even in the core area (e.g. `src/jmri/util/usb/UsbBrowserPanel.form`) I am not sure about project's policy on including such generated UI. 

Please advise.


Sorry, force pushed because the initial commit was done with a wrong identity; sorry for the confusion.

bobjacobsen commented:
Thanks!  

There's no problem AFAIK about using NetBeans to create something.  Having the .form files present allows somebody in the future to re-edit that using NetBeans.  But we don't require anybody to use NetBeans, so that code might also get edited some other way in the future, in which case the .form becomes obsolete.  

The current Travis CI run is reporting this:
```
Rule violated for source file jmri/util/ValidatingInputPane.java: lines covered count is 0, but expected minimum is 1
```

That's saying that there are no tests for the ValidatingInputPane class.  Could you please add some?  There's more info on our [Unit test page](http://jmri.org/help/en/html/doc/Technical/JUnit.shtml). Thanks.

svatoun commented:
@bobjacobsen yes, will do. I noticed that late night. Need first to setup JUnit debugging, something is screwed in the `nbproject`. Which version of NetBeans people already involved generally use - can I use 9.0 ?

bobjacobsen commented:
Not sure how many use NetBeans.  

From the command line, `ant` automatically hooks up to JUnit to run the tests. Our usual NetBeans relies on that. No separate setup needed. 

http://jmri.org/help/en/html/doc/Technical/NetBeans.shtml

danielb987 commented:
I use NetBeans 8.2 on Windows 7 and Windows 8.1 and it works fine for me.

svatoun commented:
I've added decent tests for the new class. And:
- since I've noticed that there's no support to disable the whole testsuite, I've added an `AssumeGUIRule`. Maybe this rule should be placed into another package ?
- I was able to run tests from within NetBeans IDE, but not debug (the Debug item was grayed out). I've added bindings from the `debug-single-test` action to a new target in the IDE specific file. It required to modify the main project file, in a backward compatible way. Please review carefullly.

devel-bobm commented:
Restarted Travis Headless=True Static=True account unrelated problem and associated fix in #6605 .

bobjacobsen commented:
> 	• since I've noticed that there's no support to disable the whole testsuite, I've added an AssumeGUIRule. Maybe this rule should be placed into another package ?

Don’t understand what you’re saying here, sorry. We usually have tests individually check for headless operation (because we can run them one at a time).  Then we have test target(s) that set headless, even if running with graphics.  Is AssumeGUIRule separate from that?

pabender commented:
> • since I've noticed that there's no support to disable the whole testsuite, I've added an AssumeGUIRule. Maybe this rule should be placed into another package ?
> Don’t understand what you’re saying here, sorry. We usually have tests individually check for headless operation (because we can run them one at a time). Then we have test target(s) that set headless, even if running with graphics. Is AssumeGUIRule separate from that?

I happen to agreee with Bob here.

Also it is possible that you currently have only tests that require graphics, but a test may later be added that can be run in a headless fashion.  I fact, JPanel objects can be created without being visible, so they can be created in headless mode.

I see several tests in 
https://github.com/JMRI/JMRI/pull/6596/files#diff-f98c9d794c6392a71d89cdcfe8c63f62  That may not require graphics to run.

pabender commented on java/test/jmri/util/ValidatingInputPaneTest.java:
```diff
@@ -0,0 +1,217 @@
+import static junit.framework.TestCase.assertFalse;
+import static junit.framework.TestCase.assertNotNull;
+import static junit.framework.TestCase.assertNull;
+import static junit.framework.TestCase.assertTrue;
```
You should be using .org.junit.Assert.  We should see anything in JUnit4 tests from junit.framework.

rhwood commented:
NetBeans can run tests on individual classes under `Sources` or `Unit Tests` in the Project tree by selecting the class, right clicking on it, and selecting "Test File". To test a package, find it in `Unit Tests` and run "Test File" on the file PackageTest.

svatoun commented:
re. rule: I thought it is a way how to disable all tests in the class (like the one in this PR) without having each test method individually to `assumeFalse(GraphicsEnvironment.isHeadless)`. Of course it can be seen redundant, but if the whole test class is designed to be run just in graphics mode, it's often better to declare it explicitly.

svatoun commented:
re. running tests: I wrote about **debuging** individual tests, not running. If you attempt to debug the test class - right click the test class itself, choose context menu -> Debug, or context menu -> Debug File from the Editor window, this infobox appears: http://bordel.klfree.net/~belgarat/debug-test.jpg  (in NB9.0)
As I wrote, plain Run, Run Test File (without attaching a debugger) is fine.  I still need to find out why "Debug Test File" context menu is disabled.

pabender commented:
Sent from my iPad

> On Feb 28, 2019, at 12:38 AM, Svata Dedic <notifications@github.com> wrote:
> re. rule: I thought it is a way how to disable all tests in the class (like the one in this PR) without having each test method individually to assumeFalse(GraphicsEnvironment.isHeadless). Of course it can be seen redundant, but if the whole test class is designed to be run just in graphics mode, it's often better to declare it explicitly.
> 

Our goal is actually to use the assumption on isHeadless only on the tests that absolutely require it.

As I said earlier, you have a test file in this PR that tests a JPanel, and while the whole thing tests a GUI object, there are at least a couple of tests you may be able to run without the headless assumption.  We want those tests to run whenever possible.

As an example, I don’t see why the testInconvertibleInputFails  test requires headed mode.

Paul

svatoun commented:
Strictly speaking, even the `testInconvertibleInputFails` not only constructs Swing components, but during the execution (in the tested class) makes a call to `setVisible()`. But I see your point; the tests do not actually fail in headless mode.

svatoun commented:
Fixed imports, removed unwanted Rule (only certain tests `assume` GUI). Removed leftover commented-out line in project.xml

svatoun commented:
Please advise with the `appveyor` test. It reports that 

> Rule violated for source file jmri/util/ValidatingInputPane.java: lines covered count is 0, but expected minimum is 1

but there should be at least some linest covered by the `ValidatingInputPaneTest` in the same package.

danielb987 commented:
You need to add `ValidatingInputPaneTest.class` to [jmri.util.PackageTest](https://github.com/JMRI/JMRI/blob/master/java/test/jmri/util/PackageTest.java)

svatoun commented:
@danielb987 Thanks for the patience, I should better RTM.

svatoun commented:
Next question: The message in 

> 1. ERROR in /home/travis/build/JMRI/JMRI/java/test/jmri/util/ValidatingInputPaneTest.java (at line 150)
	return (T)result[0];

seems to come from ECJ. In this particular case I'll remove the output support (it wasn't needed actually, so dead code). But how do you recommend to satisfy the check in general case: one cannot `new`-allocate an array of generic type and there are other situations where unchecked cast is necessary or at least pragmatic. Is @SuppressWarnings accepted for such cases ?

danielb987 commented:
> @danielb987 Thanks for the patience, I should better RTM.

You are welcome. I have done the same misstake several times.

danielb987 commented:
> Next question: The message in
> 
> > 1. ERROR in /home/travis/build/JMRI/JMRI/java/test/jmri/util/ValidatingInputPaneTest.java (at line 150)
> >    return (T)result[0];
> 
> seems to come from ECJ. In this particular case I'll remove the output support (it wasn't needed actually, so dead code). But how do you recommend to satisfy the check in general case: one cannot `new`-allocate an array of generic type and there are other situations where unchecked cast is necessary or at least pragmatic. Is @SuppressWarnings accepted for such cases ?

Why is `out` and `result` arrays? If there is only one element in the array, I don't see any reason for it to be an array.

svatoun commented:
@danielb987 Both `out` and `result` values are filled from a lambda passed into `invokeAndWait`. variables. Using arrays to get values out from innerclass to local variables used to be an usual idiom, if one does not want to use fields or `AtomicXXX` references.

danielb987 commented:
I think you can use `AtomicReference` instead.

https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/atomic/AtomicReference.html

geowar1 commented:
Just a quick FYI: There are several other uses of QuickPromptUtil.promptForInt in TrackSegment and LayoutShape that should probably have the predicate added to disallow invalid input values. The predicate parameter should probably also be added to the other QuickPromptUtil.promptFor*** methods so that input validation could be added to those also. You may do that here or delegate it to me and I'll do it after this PR merges. Your call. ;-)

geowar1 commented:
Also, to avoid redundant code you could have the original parseForInt return the new method passing null for the predicate:
`

    static public int promptForInt(Component parentComponent, String message, String title, int oldValue) {
        return promptForInt(parentComponent, message, title, oldValue, null);
    }
`

svatoun commented:
@geowar1: I wanted first to push this PR through the process to e.g. see whether such a change would be accepted. I'd make search & update for similar constructions as the next step.

re. redundant code: Agreed that the code should delegate eventually; I however tend to be conservative - not to damage too many clients in one shot. I'd feel safer if the new code sits in the master for some time & is used before redirecting old clients to it. I am paranoid :)

geowar1 commented:
Perfectly ok answer. If after this you want to delegate anything to me just let me know and I'll jump in.
(Adding validator predicates seem obvious in hindsight… wish I'd thought of them when I wrote that code… ;-)
